File: retrocpu_emu/src/ioboard/hex_keyboard/io_console.ts
/**
 * IO ボード前面パネル（ioboard.mdc ファンクションキー）
 *
 * キーボードからのメモリ R/W はハンドシェイク（13h/14h）。
 * 実行は 12h。表示はパネル自身が 7セグを駆動（モニタは 0x16 を使わない）。
 */

import { wordToSegDigits, wordToSegDigitsPadded } from "../seven_led/seg_font";
import {
  OFFSETS,
  alignAddrToStep,
  normalizeAddrStep,
} from "../setting_area";
import { applyLedDisplayCommand } from "../seven_led/io_led";
import {
  applyUndefLedCommand,
  getUndefLed,
  resetUndefLed,
} from "../bullet_led/io_undef_led";
import { CMD_IO_TO_CPU } from "../../shared/board_link";
import { FN_KEY_LABELS } from "../../shared/fn_keys";
import { getLogger } from "../../log/logger";

const log = getLogger("panel");
const SEG_DASH = 0x40;

export type ConsoleFocus = "addr" | "data";
export type ConsoleMode = "monitor" | "setting_area";

export type ConsoleFnKey =
  | "F0"
  | "F1"
  | "F2"
  | "F3"
  | "F4"
  | "F5"
  | "F6"
  | "F7";

export { FN_KEY_LABELS };

/** CPU ボードへのハンドシェイク／制御（IO Worker が実装） */
export type ConsoleCpuBridge = {
  /** ハンドシェイク MEM_READ (0x13): ワードアドレスから 1 ワード読む */
  memReadWord(wordAddr: number): Promise<number>;
  /** ハンドシェイク MEM_WRITE (0x14): 1 ワード書く */
  memWriteWord(wordAddr: number, word: number): Promise<void>;
  /** 設定エリア（00h-FFh）から 1 バイト読む */
  readSettingByte(byteAddr: number): Promise<number>;
  /** 設定エリア（00h-FFh）へ 1 バイト書く */
  writeSettingByte(byteAddr: number, value: number): Promise<void>;
  /** ハンドシェイク EXEC (0x12) */
  exec(wordAddr: number): Promise<void>;
  setHalt(halt: boolean): Promise<void>;
  /** F7 RST: HALT → ブートモニタ DMA → CPU RST パルス */
  pulseReset(): Promise<void>;
  /** 現在 HALT 相当か（表示用） */
  isHalted(): boolean;
};

export type IoConsoleState = {
  /** ワードアドレス（ADDR 8 桁表示） */
  wordAddr: number;
  /** データワード（DATA 4 桁） */
  dataWord: number;
  focus: ConsoleFocus;
  mode: ConsoleMode;
  halted: boolean;
  /** 未定義命令検出（IISR ビット15=LSB 0x0001 / 砲弾 B） */
  undefInsn: boolean;
};

export class IoConsole {
  private wordAddr = 0;
  private dataWord = 0;
  private focus: ConsoleFocus = "addr";
  private mode: ConsoleMode = "monitor";
  private halted = true;
  /** モニター時のアドレス増加数（設定 05h。設定エリア編集中は常に 1） */
  private addrStep = 1;
  /** モニター時の ADDR 7セグ桁数（設定 0Ah、最大 8） */
  private addrDigits = 5;
  /** モニター時の DATA 7セグ桁数（設定 0Bh、最大 4） */
  private dataDigits = 4;

  /**
   * @param cpu CPU ボードへのハンドシェイク／制御ブリッジ
   */
  constructor(private readonly cpu: ConsoleCpuBridge) {
    this.refreshLeds();
  }

  /**
   * パネルの現在状態を返す（スナップショット用）。
   * @returns アドレス／データ／入力フォーカス／HALT／未定義命令フラグ
   */
  getState(): IoConsoleState {
    return {
      wordAddr: this.wordAddr >>> 0,
      dataWord: this.dataWord & 0xffff,
      focus: this.focus,
      mode: this.mode,
      halted: this.halted,
      undefInsn: getUndefLed(),
    };
  }

  /**
   * 1階 RST 後のパネル表示（ioboard.mdc）。
   * ADDR 入力・ADDR/DATA=0・UNDEF 消灯。ブートモニタは H 待ちなので HALT。
   */
  notifyCpuReset(): void {
    this.wordAddr = 0;
    this.dataWord = 0;
    this.focus = "addr";
    this.mode = "monitor";
    resetUndefLed();
    this.halted = true;
    this.refreshLeds();
  }

  /** ADS 長押しで設定エリア編集モードへ入退する。 */
  onAdsLongPress(): void {
    this.mode = this.mode === "monitor" ? "setting_area" : "monitor";
    this.focus = "addr";
    this.wordAddr = 0;
    this.dataWord = 0;
    log.info("ADS 長押しでモード切替", { mode: this.mode });
    this.refreshLeds();
  }

  /**
   * ハンドシェイク 13h で UNDEF LED（砲弾 B）を明示設定する。
   * sticky: 消灯指示または RST まで点灯を維持する。
   * @param on true=点灯 / false=消灯
   */
  setUndefLed(on: boolean): void {
    const changed = on !== getUndefLed();
    applyUndefLedCommand(on);
    if (on) this.halted = true;
    if (changed && on) log.warn("未定義命令LED点灯（13h）");
    this.refreshLeds();
  }

  /** 16進キー 0–F */
  onHex(digit: string): void {
    const n = parseInt(digit, 16);
    if (Number.isNaN(n)) return;
    if (this.focus === "addr") {
      if (this.mode === "setting_area") {
        this.wordAddr = ((this.wordAddr << 4) | (n & 0xf)) & 0xff;
      } else {
        this.wordAddr = ((this.wordAddr << 4) | (n & 0xf)) >>> 0;
      }
    } else {
      if (this.mode === "setting_area") {
        this.dataWord = ((this.dataWord << 4) | (n & 0xf)) & 0xff;
      } else {
        this.dataWord = ((this.dataWord << 4) | (n & 0xf)) & 0xffff;
      }
    }
    log.debug("16進キー入力", {
      digit,
      focus: this.focus,
      mode: this.mode,
      addr: this.wordAddr,
      data: this.dataWord,
    });
    this.refreshLeds();
  }

  /**
   * ファンクションキーを処理する（ioboard.mdc の F0〜F7）。
   * メモリ R/W はハンドシェイク 13h/14h、実行は 12h を使う。
   * @param fn "F0"=ADS "F1"=CLR "F2"=INC "F3"=DEC "F4"=WINC "F5"=RUN "F6"=H/ST "F7"=RST
   */
  async onFunction(fn: ConsoleFnKey): Promise<void> {
    await this.syncFromSettings();
    log.info("ファンクションキー", {
      fn,
      key: FN_KEY_LABELS[fn],
      addr: this.wordAddr,
      data: this.dataWord,
    });
    switch (fn) {
      case "F0": // ADS
        this.focus = this.focus === "addr" ? "data" : "addr";
        if (this.focus === "data") {
          this.alignMonitorAddr();
          await this.readAt(this.wordAddr);
        }
        this.refreshLeds();
        break;
      case "F1": // CLR
        if (this.focus === "addr") this.wordAddr = 0;
        else this.dataWord = 0;
        this.refreshLeds();
        break;
      case "F2": // INC
        this.alignMonitorAddr();
        this.wordAddr = this.addAddr(this.addrDelta());
        await this.readAt(this.wordAddr);
        this.focus = "data";
        this.refreshLeds();
        break;
      case "F3": // DEC
        this.alignMonitorAddr();
        this.wordAddr = this.addAddr(-this.addrDelta());
        await this.readAt(this.wordAddr);
        this.focus = "data";
        this.refreshLeds();
        break;
      case "F4": // WINC
        this.alignMonitorAddr();
        if (this.mode === "setting_area") {
          await this.cpu.writeSettingByte(
            this.wordAddr & 0xff,
            this.dataWord & 0xff,
          );
        } else {
          await this.cpu.memWriteWord(this.wordAddr, this.dataWord);
        }
        this.wordAddr = this.addAddr(this.addrDelta());
        await this.readAt(this.wordAddr);
        this.focus = "data";
        this.refreshLeds();
        break;
      case "F5": // RUN
        this.alignMonitorAddr();
        await this.cpu.exec(this.wordAddr);
        this.halted = false;
        this.refreshLeds();
        break;
      case "F6": // H/ST（パネル状態でトグル。isHalted は即時反映されない／H 即停止でも RUN 表示と食い違う）
        if (this.halted) {
          await this.cpu.setHalt(false);
          this.halted = false;
        } else {
          await this.cpu.setHalt(true);
          this.halted = true;
        }
        this.refreshLeds();
        break;
      case "F7": // RST: HALT → ブートモニタ DMA → CPU RST（bridge がフルリセット）
        await this.cpu.pulseReset();
        this.notifyCpuReset();
        break;
      default:
        break;
    }
  }

  /**
   * 設定エリアからモニター用の増加数・7セグ桁数を取り込む。
   */
  private async syncFromSettings(): Promise<void> {
    this.addrStep = normalizeAddrStep(
      await this.cpu.readSettingByte(OFFSETS.ADDR_STEP),
    );
    const addrDigits = await this.cpu.readSettingByte(
      OFFSETS.SEVEN_SEG_ADDR_DIGITS,
    );
    const dataDigits = await this.cpu.readSettingByte(
      OFFSETS.SEVEN_SEG_DATA_DIGITS,
    );
    this.addrDigits =
      addrDigits >= 1 && addrDigits <= 8 ? addrDigits : 5;
    this.dataDigits =
      dataDigits >= 1 && dataDigits <= 4 ? dataDigits : 4;
  }

  /**
   * モニター時、増加数 2 なら奇数アドレスを 1 減算する。
   */
  private alignMonitorAddr(): void {
    if (this.mode !== "monitor") return;
    this.wordAddr = alignAddrToStep(this.wordAddr, this.addrStep);
  }

  /**
   * INC/DEC/WINC の増減幅。設定エリア編集中は常に 1。
   * @returns 1 または 2
   */
  private addrDelta(): number {
    return this.mode === "setting_area" ? 1 : this.addrStep;
  }

  /**
   * アドレスに増減を加える（設定エリアは 8bit ラップ）。
   * @param delta 加算する値（負数可）
   * @returns 更新後アドレス
   */
  private addAddr(delta: number): number {
    if (this.mode === "setting_area") {
      return (this.wordAddr + delta) & 0xff;
    }
    return (this.wordAddr + delta) >>> 0;
  }

  /**
   * 指定アドレスを読んで DATA 表示用の値に取り込む。
   * @param wordAddr 読み出すワードアドレス
   */
  private async readAt(wordAddr: number): Promise<void> {
    if (this.mode === "setting_area") {
      this.dataWord = (await this.cpu.readSettingByte(wordAddr & 0xff)) & 0xff;
      return;
    }
    this.dataWord = (await this.cpu.memReadWord(wordAddr)) & 0xffff;
  }

  /** ADDR/DATA + ADS(E/F) + HALT(D)/RUN(C) + UNDEF(B) */
  refreshLeds(): void {
    const addrSegs =
      this.mode === "setting_area"
        ? [SEG_DASH, 0, 0, 0, 0, 0, ...wordToSegDigits(this.wordAddr & 0xff, 2)]
        : wordToSegDigitsPadded(this.wordAddr, this.addrDigits, 8);
    const dataSegs =
      this.mode === "setting_area"
        ? [0, 0, ...wordToSegDigits(this.dataWord & 0xff, 2)]
        : wordToSegDigitsPadded(this.dataWord, this.dataDigits, 4);
    const sevenSeg = new Uint8Array(12);
    sevenSeg.set(addrSegs, 0);
    sevenSeg.set(dataSegs, 8);
    let bullet8_F = 0;
    if (this.focus === "addr")
      bullet8_F |= 1 << 6; // E
    else bullet8_F |= 1 << 7; // F
    if (this.halted)
      bullet8_F |= 1 << 5; // D = HALT
    else bullet8_F |= 1 << 4; // C = RUN
    if (getUndefLed()) bullet8_F |= 1 << 3; // B = UNDEF
    applyLedDisplayCommand({
      sevenSeg,
      bulletLed0_7: 0,
      bulletLed8_F: bullet8_F,
    });
  }
}

/** ブリッジ実装が使うコマンド定数の再エクスポート */
export const CONSOLE_HSHK = {
  MEM_READ: CMD_IO_TO_CPU.MEM_READ,
  MEM_WRITE: CMD_IO_TO_CPU.MEM_WRITE,
  EXEC: CMD_IO_TO_CPU.EXEC,
} as const;

File: retrocpu_emu_rs/src/ioboard/monitor/io_console.rs
//! IO ボード前面パネル（ioboard.mdc ファンクションキー）。
//!
//! メモリ R/W はハンドシェイク `83h`/`84h`、実行は `82h`。
//! 表示はパネル自身が 7セグ／砲弾を駆動（モニタは `16h` を使わない）。

use crate::board_link::{BoardLinkError, PanelHost};
use crate::ioboard::input::{apply_hex_digit_to_addr, apply_hex_digit_to_data};
use crate::ioboard::monitor::IoMonitor;
use crate::ioboard::output::bullet_led::{
	BulletLed, LED_ADDR, LED_DATA, LED_HALT, LED_RUN, LED_UNDEF,
};
use crate::ioboard::output::seven_segment_led::{
	word_to_seg_digits, word_to_seg_digits_padded, SevenSegmentLed, ADDR_DIGIT_COUNT,
	DATA_DIGIT_COUNT, DIGIT_COUNT, SEG_DASH,
};
use crate::ioboard::setting_area::{
	default_settings_for_cpu, normalize_addr_step, offsets, IoBoardSettings,
};

/// ADDR / DATA 入力フォーカス。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ConsoleFocus {
	/// アドレス部。
	Addr,
	/// データ部。
	Data,
}

/// モニター／設定エリア編集。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ConsoleMode {
	/// 通常モニター。
	Monitor,
	/// ADS 長押しの設定エリア。
	SettingArea,
}

/// ファンクションキー F0–F7。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ConsoleFnKey {
	/// ADS。
	F0,
	/// CLR。
	F1,
	/// INC。
	F2,
	/// DEC。
	F3,
	/// WINC。
	F4,
	/// RUN。
	F5,
	/// H/ST。
	F6,
	/// RST。
	F7,
}

impl ConsoleFnKey {
	/// `"F0"`–`"F7"` から変換。
	///
	/// # Arguments
	/// - `name`: 関数に渡す値
	///
	/// # Returns
	/// - `Option<Self>` を返します。
	pub fn from_name(name: &str) -> Option<Self> {
		match name.trim().to_ascii_uppercase().as_str() {
			"F0" => Some(Self::F0),
			"F1" => Some(Self::F1),
			"F2" => Some(Self::F2),
			"F3" => Some(Self::F3),
			"F4" => Some(Self::F4),
			"F5" => Some(Self::F5),
			"F6" => Some(Self::F6),
			"F7" => Some(Self::F7),
			_ => None,
		}
	}

	/// 表示ラベル。
	///
	/// # Returns
	/// - `&'static str` を返します。
	pub fn label(self) -> &'static str {
		match self {
			Self::F0 => "ADS",
			Self::F1 => "CLR",
			Self::F2 => "INC",
			Self::F3 => "DEC",
			Self::F4 => "WINC",
			Self::F5 => "RUN",
			Self::F6 => "H/ST",
			Self::F7 => "RST",
		}
	}
}

/// パネル状態スナップショット。
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct IoConsoleState {
	/// ワードアドレス（ADDR 表示）。
	pub word_addr: u32,
	/// データワード（DATA 表示）。
	pub data_word: u16,
	/// 入力フォーカス。
	pub focus: ConsoleFocus,
	/// モード。
	pub mode: ConsoleMode,
	/// HALT 表示。
	pub halted: bool,
	/// UNDEF（砲弾 B）。
	pub undef_insn: bool,
}

/// IO ボード前面コンソール。
#[derive(Debug, Clone)]
pub struct IoConsole {
	word_addr: u32,
	data_word: u16,
	focus: ConsoleFocus,
	mode: ConsoleMode,
	halted: bool,
	undef_insn: bool,
	addr_step: u8,
	addr_digits: u8,
	data_digits: u8,
}

impl Default for IoConsole {
	fn default() -> Self {
		Self::new(&IoBoardSettings::default())
	}
}

impl IoConsole {
	/// 設定から桁数・増加数を初期化する。
	///
	/// # Arguments
	/// - `settings`: 設定値
	///
	/// # Returns
	/// - 初期化済みインスタンスを返します。
	pub fn new(settings: &IoBoardSettings) -> Self {
		Self {
			word_addr: 0,
			data_word: 0,
			focus: ConsoleFocus::Addr,
			mode: ConsoleMode::Monitor,
			halted: true,
			undef_insn: false,
			addr_step: normalize_addr_step(settings.addr_step),
			addr_digits: settings.seven_seg_addr_digits.clamp(1, 8),
			data_digits: settings.seven_seg_data_digits.clamp(1, 4),
		}
	}

	/// 現在状態。
	///
	/// # Returns
	/// - `IoConsoleState` を返します。
	pub fn state(&self) -> IoConsoleState {
		IoConsoleState {
			word_addr: self.word_addr,
			data_word: self.data_word,
			focus: self.focus,
			mode: self.mode,
			halted: self.halted,
			undef_insn: self.undef_insn,
		}
	}

	/// RST 後のパネル表示（ADDR 入力・値 0・UNDEF 消灯・HALT）。
	pub fn notify_cpu_reset<H: PanelHost>(
		&mut self,
		host: &mut H,
		seven: &mut SevenSegmentLed,
		bullet: &mut BulletLed,
	) {
		self.sync_from_settings(host);
		self.word_addr = 0;
		self.data_word = 0;
		self.focus = ConsoleFocus::Addr;
		self.mode = ConsoleMode::Monitor;
		self.undef_insn = false;
		self.halted = true;
		self.refresh_leds(seven, bullet);
	}

	/// ADS 長押しで設定エリア編集モードへ入退する。
	pub fn on_ads_long_press<H: PanelHost>(
		&mut self,
		host: &mut H,
		seven: &mut SevenSegmentLed,
		bullet: &mut BulletLed,
	) {
		let leaving = self.mode == ConsoleMode::SettingArea;
		self.mode = if self.mode == ConsoleMode::Monitor {
			ConsoleMode::SettingArea
		} else {
			ConsoleMode::Monitor
		};
		self.focus = ConsoleFocus::Addr;
		self.word_addr = 0;
		self.data_word = 0;
		if leaving {
			self.sync_from_settings(host);
		}
		self.refresh_leds(seven, bullet);
	}

	/// CPU 実行状態を砲弾 C/D に反映する。
	pub fn sync_cpu_halted(
		&mut self,
		halted: bool,
		seven: &mut SevenSegmentLed,
		bullet: &mut BulletLed,
	) {
		if self.halted == halted {
			return;
		}
		self.halted = halted;
		self.refresh_leds(seven, bullet);
	}

	/// ハンドシェイク `13h` 相当の UNDEF LED。
	pub fn set_undef_led(&mut self, on: bool, seven: &mut SevenSegmentLed, bullet: &mut BulletLed) {
		self.undef_insn = on;
		if on {
			self.halted = true;
		}
		self.refresh_leds(seven, bullet);
	}

	/// 16 進キー 0–F（1 ニブル）。
	pub fn on_hex(&mut self, digit: u8, seven: &mut SevenSegmentLed, bullet: &mut BulletLed) {
		match self.focus {
			ConsoleFocus::Addr => {
				self.word_addr =
					apply_hex_digit_to_addr(self.word_addr, digit, self.mode == ConsoleMode::SettingArea);
			}
			ConsoleFocus::Data => {
				self.data_word =
					apply_hex_digit_to_data(self.data_word, digit, self.mode == ConsoleMode::SettingArea);
			}
		}
		self.refresh_leds(seven, bullet);
	}

	/// ファンクションキー処理。
	pub fn on_function<H: PanelHost>(
		&mut self,
		fn_key: ConsoleFnKey,
		host: &mut H,
		seven: &mut SevenSegmentLed,
		bullet: &mut BulletLed,
	) -> Result<(), BoardLinkError> {
		self.sync_from_settings(host);
		match fn_key {
			ConsoleFnKey::F0 => {
				self.focus = match self.focus {
					ConsoleFocus::Addr => ConsoleFocus::Data,
					ConsoleFocus::Data => ConsoleFocus::Addr,
				};
				if self.focus == ConsoleFocus::Data {
					self.align_monitor_addr();
					self.read_at(host, self.word_addr)?;
				}
				self.refresh_leds(seven, bullet);
			}
			ConsoleFnKey::F1 => {
				if self.focus == ConsoleFocus::Addr {
					self.word_addr = 0;
				} else {
					self.data_word = 0;
				}
				self.refresh_leds(seven, bullet);
			}
			ConsoleFnKey::F2 => {
				self.align_monitor_addr();
				self.word_addr = self.add_addr(self.addr_delta() as i32);
				self.read_at(host, self.word_addr)?;
				self.focus = ConsoleFocus::Data;
				self.refresh_leds(seven, bullet);
			}
			ConsoleFnKey::F3 => {
				self.align_monitor_addr();
				self.word_addr = self.add_addr(-(self.addr_delta() as i32));
				self.read_at(host, self.word_addr)?;
				self.focus = ConsoleFocus::Data;
				self.refresh_leds(seven, bullet);
			}
			ConsoleFnKey::F4 => {
				self.align_monitor_addr();
				IoMonitor::write_word(
					host,
					self.word_addr,
					self.data_word,
					self.mode == ConsoleMode::SettingArea,
				)?;
				self.word_addr = self.add_addr(self.addr_delta() as i32);
				self.read_at(host, self.word_addr)?;
				self.focus = ConsoleFocus::Data;
				self.refresh_leds(seven, bullet);
			}
			ConsoleFnKey::F5 => {
				self.align_monitor_addr();
				IoMonitor::exec_word(host, self.word_addr)?;
				self.halted = false;
				self.refresh_leds(seven, bullet);
			}
			ConsoleFnKey::F6 => {
				if self.halted {
					host.start_run()?;
					self.halted = false;
				} else {
					host.request_halt()?;
					self.halted = true;
				}
				self.refresh_leds(seven, bullet);
			}
			ConsoleFnKey::F7 => {
				host.reset_and_reload_monitor()?;
				self.notify_cpu_reset(host, seven, bullet);
			}
		}
		Ok(())
	}

	/// 設定から増加数・桁数を取り込む。
	///
	/// # Arguments
	/// - `host`: 関数に渡す値
	pub fn sync_from_settings<H: PanelHost>(&mut self, host: &H) {
		let cpu = host.read_setting_byte(offsets::CPU_TYPE as u8);
		let defaults = default_settings_for_cpu(cpu);
		self.addr_step = normalize_addr_step(host.read_setting_byte(offsets::ADDR_STEP as u8));
		let ad = host.read_setting_byte(offsets::SEVEN_SEG_ADDR_DIGITS as u8);
		let dd = host.read_setting_byte(offsets::SEVEN_SEG_DATA_DIGITS as u8);
		self.addr_digits = if (1..=8).contains(&ad) {
			ad
		} else {
			defaults.seven_seg_addr_digits
		};
		self.data_digits = if (1..=4).contains(&dd) {
			dd
		} else {
			defaults.seven_seg_data_digits
		};
	}

	fn align_monitor_addr(&mut self) {
		if self.mode != ConsoleMode::Monitor {
			return;
		}
		self.word_addr = IoMonitor::align_word_addr(self.word_addr, self.addr_step);
	}

	fn addr_delta(&self) -> u8 {
		IoMonitor::addr_delta(self.mode == ConsoleMode::SettingArea, self.addr_step)
	}

	fn add_addr(&self, delta: i32) -> u32 {
		IoMonitor::shift_word_addr(self.word_addr, delta, self.mode == ConsoleMode::SettingArea)
	}

	fn read_at<H: PanelHost>(&mut self, host: &mut H, word_addr: u32) -> Result<(), BoardLinkError> {
		self.data_word = IoMonitor::read_word(host, word_addr, self.mode == ConsoleMode::SettingArea)?;
		Ok(())
	}

	/// ADDR/DATA + ADS(E/F) + HALT(D)/RUN(C) + UNDEF(B) を LED に載せる。
	///
	/// # Arguments
	/// - `seven`: 関数に渡す値
	/// - `bullet`: 関数に渡す値
	pub fn refresh_leds(&self, seven: &mut SevenSegmentLed, bullet: &mut BulletLed) {
		let mut patterns = [0u8; DIGIT_COUNT];
		if self.mode == ConsoleMode::SettingArea {
			let addr = word_to_seg_digits(self.word_addr & 0xff, 2);
			patterns[0] = SEG_DASH;
			patterns[6] = addr[0];
			patterns[7] = addr[1];
			let data = word_to_seg_digits(u32::from(self.data_word & 0xff), 2);
			patterns[10] = data[0];
			patterns[11] = data[1];
		} else {
			let addr =
				word_to_seg_digits_padded(self.word_addr, self.addr_digits as usize, ADDR_DIGIT_COUNT);
			let data = word_to_seg_digits_padded(
				u32::from(self.data_word),
				self.data_digits as usize,
				DATA_DIGIT_COUNT,
			);
			patterns[..ADDR_DIGIT_COUNT].copy_from_slice(&addr);
			patterns[ADDR_DIGIT_COUNT..].copy_from_slice(&data);
		}
		seven.set_patterns(&patterns);

		let mut bits: u16 = 0;
		match self.focus {
			ConsoleFocus::Addr => bits |= 1 << LED_ADDR,
			ConsoleFocus::Data => bits |= 1 << LED_DATA,
		}
		if self.halted {
			bits |= 1 << LED_HALT;
		} else {
			bits |= 1 << LED_RUN;
		}
		if self.undef_insn {
			bits |= 1 << LED_UNDEF;
		}
		// 砲弾 0–A は触らない（ユーザー／16h 用）。B–F だけ差し替え。
		let lo = bullet.bits() & 0x07ff;
		bullet.set_bits(lo | (bits & 0xf800));
	}
}
